uniffi_bindgen/src/bindings/kotlin/templates/EnumTemplate.kt

```diff
@@ -1,19 +1,94 @@
-enum class {{ e.name()|class_name_kt }} {
-    {% for variant in e.variants() %}
-    {{ variant|enum_variant_kt }}{% if loop.last %};{% else %},{% endif %}
+{#
+// Kotlin's `enum class` constuct doesn't support variants with associated data,
+// but is a little nicer for consumers than its `sealed class` enum pattern.
+// So, we switch here, using `enum class` for enums with no associated data
+// and `sealed class` for the general case.
```

Comment:
@jhugman here's a design question I'd appreciate your insight on. Kotlin's `enum class` declaration doesn't support associated data in the way we want to use it, so in the general case we have to use the `sealed class` approach. However, when there isn't any associated data, Kotlin's `enum` seems to be a little nicer to work with. Here I've opted to switch between them - enums without data will be an `enum class` just as they were under the previous implementation, while newly-supported enums with data will be a `sealed class`.

This has the nice property of being backwards-compatible for existing consumers of existing UDL files. But I'm a little concerned about potential developer confusion in future.

For example, suppose someone currently has a plain-old-enum with no data in their UDL file, and their Kotlin consumers are consuming it through an `enum class`. The component API evolves and adds a new case to the enum, this time with some associated data. For Kotlin consumers this addition will *change the entire type* of the thing they're dealing with, including changing the capitalization of existing items as it switches from being an `enum class` to a `sealed class`.

Maybe this is OK, since consumers will need to update their code to handle the new variant anyway. Or maybe it's leaking implementation details in a way that we shouldn't need to do.

Thoughts?

Reply:
It's a great question. 

From our (nimbus + uniffi) point of view, I don't think it matters: we have a tiny amount of code using UDL enums, so changing them will have some coordination cost when changing them to a new format.

For android devs: adding another `enum` variant is going to be a breaking change.

There is some JVM cost to using sealed classes over Java style (around class initialization, maximum method counts etc). I don't think we're at the stage we need to worry about that.

While thinking about this, it did strike me that _Kotlin_ has had to make a similar choice: Java enums existed after  Java5. Kotlin came up with a better way, yet still supported old school Java enums. I should imagine their backwards compatibility problem was more persuasive/pressing than ours.

Reply:
Other things this question made me think about:

 * [Typescript can do Tagged Unions][1], and whatever Javascript can support natively.
 * [Javascript has barely has the simple enums][0].

Until we have a idiomatically _consumable_ version of tagged unions in JS, we're going to be discouraging use of this feature.



[0]: https://www.andrewnater.com/posts/enums-in-javascript/
[1]: https://gregoryppabian.medium.com/creating-tagged-unions-in-typescript-ce727a306878

Reply:
> Until we have a idiomatically consumable version of tagged unions in JS, we're going to be discouraging use of this feature.

Yeah. Python has a similar problem, which is one of the reasons I like having the python backend here, to think about how things will look in a language with basically the opposite of a Rust-style strong algebraic type system.